pkg/chunk/storage/caching_storage_client.go

```diff
@@ -3,35 +3,117 @@ package storage
 import (
 	"bytes"
 	"context"
+	"encoding/gob"
+	"encoding/hex"
+	"hash/fnv"
 	"strings"
 	"time"
 
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+
 	"github.com/weaveworks/cortex/pkg/chunk"
 	"github.com/weaveworks/cortex/pkg/chunk/cache"
 )
 
+var (
+	cacheCorruptErrs = promauto.NewCounter(prometheus.CounterOpts{
+		Name: "querier_index_cache_corruptions_total",
+		Help: "The number of cache corruptions for the index cache.",
+	})
+	cacheHits = promauto.NewCounter(prometheus.CounterOpts{
+		Name: "querier_index_cache_hits_total",
+		Help: "The number of cache hits for the index cache.",
+	})
+	cacheGets = promauto.NewCounter(prometheus.CounterOpts{
+		Name: "querier_index_cache_gets_total",
+		Help: "The number of gets for the index cache.",
+	})
+	cachePuts = promauto.NewCounter(prometheus.CounterOpts{
+		Name: "querier_index_cache_puts_total",
+		Help: "The number of puts for the index cache.",
+	})
+	cacheEncodeErrs = promauto.NewCounter(prometheus.CounterOpts{
+		Name: "querier_index_cache_encode_errors_total",
+		Help: "The number of errors for the index cache while encoding the body.",
+	})
+)
+
+// IndexCache describes the cache for the Index.
+type IndexCache interface {
+	Store(ctx context.Context, key string, val readBatch)
+	Fetch(ctx context.Context, key string) (val readBatch, ok bool, err error)
+	Stop() error
+}
+
+type indexCache struct {
+	cache.Cache
+}
+
+func (c *indexCache) Store(ctx context.Context, key string, val readBatch) {
+	cachePuts.Inc()
+
+	var buf bytes.Buffer
+	if err := gob.NewEncoder(&buf).Encode(val); err != nil {
```

Comment:
Ohh, not sure using gobs is a good idea.  I'd much rather see protos used here.

Reply:
(Performance being one reason, the other being we don't use Gobs anywhere else - its json or protos.)